_posts/2021-11-01-r-6.0.0.md

```diff
@@ -0,0 +1,140 @@
+
+# Joins 
+
+Multiple Arrow tables and datasets can now be joined in queries. 
+
+Say I want to get a table of all the flights from JFK to Las Vegas Airport on 
+9th October 2013, with the full name of the airline included.
+
+```r
+arrow_table(nycflights13::flights) %>%
+  filter(
+    year == 2013,
+    month == 10,
+    day == 9,
+    origin == "JFK",
+    dest == "LAS"
+    ) %>%
+  select(dep_time, arr_time, carrier) %>%
+  left_join(
+    arrow_table(nycflights13::airlines)
+   ) %>%
+  collect()
+```
+
+# Big changes to the execution of queries under-the-hood 
+
+Both of the first two points are driven by a large under-the-hood change to the way that dplyr pipelines are constructed and executed in R. There are almost no changes (besides new capabilities) that one would run into, but the improvement unlocked grouped aggregation, joins, and much more to come!
+
+A few small but notable changes came with this:
+`summarise()` when used with Arrow tables does not eagerly evaluate, instead you must call `collect()` or `compute()` to evaluate the query. This is inline with how datasets worked before as well as how a number of other dplyr backends work.
+the order of dataset queries is no longer deterministic. If you need a stable sort order, you should use `arrange()` at the end of your query.
+
+# Integration with DuckDB
+
+If you have the [duckdb](https://duckdb.org/) package installed, you can hand off an Arrow Dataset or query object to duckdb for further querying using the `to_duckdb()` function. This allows you to use duckdb's `dbplyr` methods, as well as its SQL interface, to aggregate data. Filtering and column projection done before `to_duckdb()` is evaluated in Arrow.  You can also hand off DuckDB data (or the result of a query) to arrow with `to_arrow()`.
+
+In the example below, we are looking at flights between NYC and Chicago, and want to avoid the worst-of-the-worst delays. To do this, we can use `percent_rank()`; however that requires a window function which isn’t yet available in Arrow, so let’s try sending the data to DuckDB to do that, then pull it back into Arrow:
+
+```r
+library(arrow, warn.conflicts = FALSE)
+library(dplyr, warn.conflicts = FALSE)
+ 
+# use duckdb to do a mutate after a group_by
+flights_filtered <- arrow_table(nycflights13::flights) %>%
+  select(carrier, origin, dest, arr_delay) %>%
+  # arriving early doesn't matter, so call negative delays 0
+  mutate(arr_delay = pmax(arr_delay, 0)) %>%
+  to_duckdb() %>%
+  # for each carrier-origin-dest, take the worst 5% of delays
+  group_by(carrier, origin, dest) %>%
+  mutate(arr_delay_rank = percent_rank(arr_delay)) %>%
+  filter(arr_delay_rank > 0.95)
+
+# pull data back into arrow to complete analysis
+flights_filtered %>%
```

Comment:
Why not just make this one long dplyr pipeline?

Reply:
This might be my fault with poor drafting / description here. I intended to show the first pipeline standing on its own (i.e. "look you can go from arrow to duckdb") and then the second pipeline would be "and then back again".

We don't have to divide it in that way (and honestly that intermediary/first pipeline is not an interesting data.frame anyway, so narratively it's a bit weird) and if we do do it in one shot, I agree we should just put it into one pipeline and say something like "This one does both `to_duckdb()` and `to_arrow()` but of course you don't have to use both like this, you can use just `to_duckdb()` alone if you want"